Blank map after background location update

#1821 fixes #1460 in that background location updates no longer crash the app in the background. However, it introduces a serious regression in its stead: manually opening the app past that point causes it to launch with a blank, totally black map that you can't interact with until you switch out and back into it with the app switcher.

Opening the app apparently causes it to be revived from the suspended state, and for some reason MGLView is either failing to receive the entered-foreground notification or is bailing because the application state is still "background", not even "inactive".

This behavior is worse than before, because system would automatically relaunch the process if it crashed in the background. Now you just get a defective view.

/cc @incanus @friedbunny

I don’t know if it’s related, but whenever I attempt to simulate a background launch via Xcode’s “Launch due to background fetch event” setting (with an app that has background fetches enabled), I hit the following OpenGL error and downstream mbgl exception:

```
[ERROR] {Map}[Shader]: Vertex shader plain failed to compile: precision highp float;
attribute vec2 a_pos;

uniform mat4 u_matrix;

void main() {
    gl_Position = u_matrix * vec4(a_pos, 0, 1);
}
libc++abi.dylib: terminating with uncaught exception of type mbgl::util::ShaderException: Vertex shader plain failed to compile
```

with this stack:

```
#8	0x0000000195b81060 in std::rethrow_exception(std::exception_ptr) ()
#9	0x00000001003e1c7c in std::__1::__assoc_state<mbgl::MapContext::RenderResult>::move() at /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/future:681
#10	0x00000001003dc3ac in std::__1::future<mbgl::MapContext::RenderResult>::get() [inlined] at /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/future:1159
#11	0x00000001003dc39c in mbgl::MapContext::RenderResult mbgl::util::Thread<mbgl::MapContext>::invokeSync<mbgl::MapContext::RenderResult, mbgl::MapContext::RenderResult (mbgl::MapContext::*)(mbgl::TransformState const&, mbgl::FrameData const&), mbgl::TransformState const, mbgl::FrameData>(mbgl::MapContext::RenderResult (mbgl::MapContext::*)(mbgl::TransformState const&, mbgl::FrameData const&), mbgl::TransformState const&&, mbgl::FrameData&&) at /path/to/gl-native/src/mbgl/util/thread.hpp:51
#12	0x00000001003d9e14 in mbgl::Map::renderSync() at /path/to/gl-native/src/mbgl/map/map.cpp:58
#13	0x0000000100484714 in -[MGLMapView glkView:drawInRect:] at /path/to/gl-native/platform/ios/MGLMapView.mm:720
#14	0x00000001857c00f0 in -[GLKView _display:] ()
```

This causes the app to crash. However, in the wild, when we run into this issue, the app does not crash; it remains usable except that tiles and annotations never render.

--- a/platform/ios/MGLMapView.mm
+++ b/platform/ios/MGLMapView.mm
@@ -213,7 +213,6 @@ std::chrono::steady_clock::duration secondsAsDuration(float duration)
         [self createGLView];
     }
 
-
     self.accessibilityLabel = @"Map";
     self.backgroundColor = [UIColor clearColor];
     self.clipsToBounds = YES;
@@ -251,7 +250,7 @@ std::chrono::steady_clock::duration secondsAsDuration(float duration)
     //
     _annotationIDsByAnnotation = [NSMapTable mapTableWithKeyOptions:NSMapTableStrongMemory valueOptions:NSMapTableStrongMemory];
 
-    _annotationImages = [NSMutableDictionary new];
+    _annotationImages = [NSMutableDictionary dictionary];
 
     std::string defaultSymbolName([MGLDefaultStyleMarkerSymbolName UTF8String]);
     _mbglMap->setDefaultPointAnnotationSymbol(defaultSymbolName);
@@ -376,9 +375,7 @@ std::chrono::steady_clock::duration secondsAsDuration(float duration)
 
 - (void)createGLView
 {
-    if (_context || [UIApplication sharedApplication].applicationState == UIApplicationStateBackground) {
-        return;
-    }
+    if (_context) return;
     
     // create context
     //
@@ -413,17 +410,6 @@ std::chrono::steady_clock::duration secondsAsDuration(float duration)
 
         return reinterpret_cast<mbgl::gl::glProc>(symbol);
     });
-}
-
-- (void)viewWillEnterForeground
-{
-    [[NSNotificationCenter defaultCenter] removeObserver:self
-                                                    name:UIApplicationWillEnterForegroundNotification
-                                                  object:nil];
-    [[NSNotificationCenter defaultCenter] removeObserver:self
-                                                    name:UIApplicationDidBecomeActiveNotification
-                                                  object:nil];
-    [self commonInit];
 }
 
 - (void)reachabilityChanged:(NSNotification *)notification
@@ -797,11 +783,11 @@ std::chrono::steady_clock::duration secondsAsDuration(float duration)
 {
     MGLAssertIsMainThread();
 
-    [self createGLView];
-    if (self.isDormant)
+    if (self.isDormant && [UIApplication sharedApplication].applicationState != UIApplicationStateBackground)
     {
         self.dormant = NO;
         
+        [self createGLView];
         [MGLMapboxEvents validate];
 
         self.glSnapshotView.hidden = YES;

